File: connect4/board.py
from __future__ import annotations

import copy
from typing import Optional

ROWS = 6
COLS = 7
EMPTY = 0
P1 = 1
P2 = 2


class Board:
    def __init__(self):
        self.grid = [[EMPTY] * COLS for _ in range(ROWS)]

    def is_valid_move(self, col: int) -> bool:
        return 0 <= col < COLS and self.grid[0][col] == EMPTY

    def get_valid_moves(self) -> list[int]:
        return [c for c in range(COLS) if self.is_valid_move(c)]

    def drop_piece(self, col: int, player: int) -> int:
        if not self.is_valid_move(col):
            raise ValueError(f"Invalid move: column {col}")
        for row in range(ROWS - 1, -1, -1):
            if self.grid[row][col] == EMPTY:
                self.grid[row][col] = player
                return row
        raise ValueError(f"Column {col} is full")

    def check_winner(self) -> Optional[int]:
        for r in range(ROWS):
            for c in range(COLS):
                p = self.grid[r][c]
                if p == EMPTY:
                    continue
                # horizontal
                if c + 3 < COLS and all(self.grid[r][c + i] == p for i in range(4)):
                    return p
                # vertical
                if r + 3 < ROWS and all(self.grid[r + i][c] == p for i in range(4)):
                    return p
                # diagonal down-right
                if r + 3 < ROWS and c + 3 < COLS and all(self.grid[r + i][c + i] == p for i in range(4)):
                    return p
                # diagonal down-left
                if r + 3 < ROWS and c - 3 >= 0 and all(self.grid[r + i][c - i] == p for i in range(4)):
                    return p
        return None

    def is_draw(self) -> bool:
        return self.check_winner() is None and len(self.get_valid_moves()) == 0

    def copy(self) -> "Board":
        return copy.deepcopy(self)

    def __str__(self) -> str:
        symbols = {EMPTY: ".", P1: "X", P2: "O"}
        rows = [" ".join(symbols[cell] for cell in row) for row in self.grid]
        header = " ".join(str(c) for c in range(COLS))
        return "\n".join(rows) + "\n" + header

File: web/src/engine/board.ts
export const ROWS = 6;
export const COLS = 7;
export const EMPTY = 0;
export const P1 = 1;
export const P2 = 2;

export type Cell = 0 | 1 | 2;
export type Player = 1 | 2;
export type Winner = Player | null;

export class Board {
  readonly grid: Int8Array;

  constructor(grid?: Int8Array) {
    this.grid = grid ?? new Int8Array(ROWS * COLS);
  }

  private at(row: number, col: number): number {
    return this.grid[row * COLS + col];
  }

  private set(row: number, col: number, value: number): void {
    this.grid[row * COLS + col] = value;
  }

  isValidMove(col: number): boolean {
    return col >= 0 && col < COLS && this.at(0, col) === EMPTY;
  }

  getValidMoves(): number[] {
    const moves: number[] = [];
    for (let c = 0; c < COLS; c++) {
      if (this.isValidMove(c)) moves.push(c);
    }
    return moves;
  }

  dropPiece(col: number, player: Player): number {
    if (!this.isValidMove(col)) {
      throw new Error(`Invalid move: column ${col}`);
    }
    for (let row = ROWS - 1; row >= 0; row--) {
      if (this.at(row, col) === EMPTY) {
        this.set(row, col, player);
        return row;
      }
    }
    throw new Error(`Column ${col} is full`);
  }

  checkWinner(): Winner {
    for (let r = 0; r < ROWS; r++) {
      for (let c = 0; c < COLS; c++) {
        const p = this.at(r, c);
        if (p === EMPTY) continue;
        // horizontal
        if (
          c + 3 < COLS &&
          this.at(r, c + 1) === p &&
          this.at(r, c + 2) === p &&
          this.at(r, c + 3) === p
        ) {
          return p as Player;
        }
        // vertical
        if (
          r + 3 < ROWS &&
          this.at(r + 1, c) === p &&
          this.at(r + 2, c) === p &&
          this.at(r + 3, c) === p
        ) {
          return p as Player;
        }
        // diagonal down-right
        if (
          r + 3 < ROWS &&
          c + 3 < COLS &&
          this.at(r + 1, c + 1) === p &&
          this.at(r + 2, c + 2) === p &&
          this.at(r + 3, c + 3) === p
        ) {
          return p as Player;
        }
        // diagonal down-left
        if (
          r + 3 < ROWS &&
          c - 3 >= 0 &&
          this.at(r + 1, c - 1) === p &&
          this.at(r + 2, c - 2) === p &&
          this.at(r + 3, c - 3) === p
        ) {
          return p as Player;
        }
      }
    }
    return null;
  }

  isDraw(): boolean {
    return this.checkWinner() === null && this.getValidMoves().length === 0;
  }

  copy(): Board {
    return new Board(new Int8Array(this.grid));
  }

  get(row: number, col: number): Cell {
    return this.at(row, col) as Cell;
  }

  toString(): string {
    const symbols = [".", "X", "O"];
    const lines: string[] = [];
    for (let r = 0; r < ROWS; r++) {
      const cells: string[] = [];
      for (let c = 0; c < COLS; c++) {
        cells.push(symbols[this.at(r, c)]);
      }
      lines.push(cells.join(" "));
    }
    const header = Array.from({ length: COLS }, (_, i) => String(i)).join(" ");
    return lines.join("\n") + "\n" + header;
  }
}
